compiler/qsc_passes/src/measurement.rs

```diff
@@ -0,0 +1,127 @@
+    }
+    errors
+}
+
+fn validate_measurement_declaration(decl: &CallableDecl, attrs: &[Attr], errors: &mut Vec<Error>) {
+    // 1. Check that the declaration takes at least one argument.
+    if decl.input.ty == Ty::UNIT {
+        errors.push(Error::NoArguments(decl.input.span));
+    }
+
+    // 2. Check that the declaration only takes Qubits as arguments.
+    match &decl.input.ty {
+        Ty::Prim(Prim::Qubit) => (),
+        Ty::Tuple(types) => {
+            for ty in types {
+                if !matches!(ty, Ty::Prim(Prim::Qubit)) {
+                    errors.push(Error::NonQubitArgument(decl.input.span));
+                    // break so that we don't repeat the same error multiple times
+                    break;
+                }
+            }
+        }
+        _ => errors.push(Error::NonQubitArgument(decl.input.span)),
+    }
+
+    // 3. Check that the declaration only outputs Results: specifically check for Unit.
+    if decl.output == Ty::UNIT {
+        errors.push(Error::NonResultOutput(decl.span));
+    }
+
+    // 3. Check that the declaration only outputs Results.
+    match &decl.output {
+        Ty::Prim(Prim::Result) => (),
+        Ty::Tuple(types) => {
+            for ty in types {
+                if !matches!(ty, Ty::Prim(Prim::Result)) {
+                    errors.push(Error::NonResultOutput(decl.span));
+                    // break so that we don't repeat the same error multiple times
+                    break;
+                }
+            }
+        }
+        _ => errors.push(Error::NonResultOutput(decl.span)),
+    }
+
+    // 4. Check that the declaration is an intrinsic.
+    if !decl_is_intrinsic(decl, attrs) {
+        errors.push(Error::NotIntrinsic(decl.span));
+    }
+}
+
+/// Returns `true` if a declaration is an intrinsic. A declaration is
+/// an intrinsic if it has `body intrinsic;` in its body or if it has
+/// the `@SimulatableIntrinsic()` attribute.
+fn decl_is_intrinsic(decl: &CallableDecl, attrs: &[Attr]) -> bool {
+    matches!(decl.body.body, SpecBody::Gen(SpecGen::Intrinsic))
+        || attrs
+            .iter()
+            .any(|attr| matches!(attr, Attr::SimulatableIntrinsic))
+}
+
+/// Returns a list of all the measurement callables and their attributes in a given `Package`.
+fn get_measurements(package: &Package) -> Vec<(&CallableDecl, &[Attr])> {
+    let mut finder = MeasurementFinder {
+        callables: Vec::new(),
+    };
+    finder.visit_package(package);
+    finder.callables
+}
+
+/// A helper structure to find the measurement callables in a Package.
+struct MeasurementFinder<'a> {
+    callables: Vec<(&'a CallableDecl, &'a [Attr])>,
+}
+
+impl<'a> Visitor<'a> for MeasurementFinder<'a> {
+    fn visit_item(&mut self, item: &'a Item) {
+        if let ItemKind::Callable(callable) = &item.kind {
+            if item.attrs.contains(&Attr::Measurement) {
+                self.callables.push((callable, &item.attrs));
```

Comment:
Rather than collecting all of the callables and then checking them, you could call the `validate_measurement_declaration` right in here and collect the errors. That technically scales better here, though I doubt we'll end up in a situation with thousands of these such that the difference would be noticeable :).

Reply:
Good catch. I changed the code to do that instead.